### ISSUE TITLE
use external Pd object

### ISSUE THREAD

**Issue by debeul** [pierreguillot/Camomile#138]
environment**
    - **OS**: MacOS 10.11.5
    - **DAW**:  Ableton Live 9 - 64 bits,
    - **Plugin**: VST2/ and AU 
    - **Audio**: for example 44100 Hz - 64 samples 
   
  Hello. And congratulations for this great plug in ... :)
I try to make a plug in to use the external object {bassemu ~} in live. But no sound comes out ... Is it possible to use external to make plug in? And if so, is there a specific approach to follow? Thank you
Axel

**Comment by alfonso73** [comment]
Hi Pierre,
I wonder if the following will be interesting for you about loading a dll
dynamically in Camomile VST.

There's is a port of VCV rack wich is called VeeSeeVST. VCV rack is a
modular framework that loads dynamically modules that are dll.
At the beginning the VeeSeeVST port coder can't resolve the dynamic loading
but then he found a way.

You may ask him some hints?

https://www.kvraudio.com/forum/viewtopic.php?f=23&t=507216&sid=d79eab49cbf68184faa0d08421f5d64e

The project is on github to.

https://github.com/bsp2/VeeSeeVSTRack

best and thanks for the great Camomile and faust in pd!



Il ven 28 set 2018, 22:22 Pierre Guillot <notifications@github.com> ha
scritto:

**Comment by alfonso73** [comment]
This seems to be the commit dealing with dynamically loaded dll

https://github.com/bsp2/VeeSeeVSTRack/commit/7e79bd3eda995f4c91126a676c87766cdafac4ff


Il ven 28 set 2018, 23:41 alfonso santimone <alfonso.santimone@gmail.com>
ha scritto:

**Comment by debeul** [comment]
thanks for the answer and sorry for not having read everything before :)

**Comment by alfonso73** [comment]
Hi @pierreguillot , is there any advancement in incorporating external libraries in Camomile? Thanks and all the best!
a.

**Comment by pierreguillot** [comment]
The only solution I found so far is to integrate the library to Camomile at compile-time. And even if I manage to load dynamically an external, it must be compiled with the multi-instance and multi-thread support so you would never be able to use the binaries used in Pd. It means that devs will have to provide a specific version of their libraries.

**Comment by pierreguillot** [comment]
FYI, I made a [quick video](https://nubo.ircam.fr/index.php/s/Z6drYf9s87e6bD6
) that explains how to compile Camomile with external libraries on macOS (in french).

**Comment by alfonso73** [comment]
Thanks @pierreguillot (btw even if my french is terrible, there's no audio in your video). But everything seems very clear and i'll try to do the same on Juce on my Win10 with Visual Studio. Any info i can find about multi-instance and multi-thread support in pd externals? It seems to be a pretty recent thing...

**Comment by alfonso73** [comment]
Thanks Pierre,
i'll try to reproduce the steps on VS2019.
The only thing that i don't understand very well in a VS perspective is
when you disable the "AU" Target Membership (at 02:14 in your video). I'll
try to find the same thing.
BTW as i understand if i compile at compile time the external libraries
they are not needed to be coded with multi-thread, multi-instance support
because libpd is providing this feature straight away.
Am i right?
thanks!


www.elgallorojorecords.bandcamp.com/
<https://elgallorojorecords.bandcamp.com/>
soundcloud.com/alfonsosantimone
www.facebook.com/alfonsosantimone

**Comment by pierreguillot** [comment]
There is no need to disable the AU target membership as there is no AU on Windows. Indeed, if you compile the externals with libpd, the externals should have the multi-instance/thread support.

**Comment by alfonso73** [comment]
Thanks Pierre! I wonder if your faustgen~ could be compiled with Camomile.
In the next period i'll try at least with zexy, cyclone, else and ceammc.
all the best.

**Comment by pierreguillot** [comment]
Yes, of course, you can compile Camomile with faustgen~, but as the faustgen~ compilation is complicated (you must link against LLVM and faust libs), the integation with camomile will be more complicated.

**Comment by pierreguillot** [comment]
Here is another the [video](https://nubo.ircam.fr/index.php/s/ozZi9CrQPLQ2dyW) for Windows that shows how to compile Camomile with externals. I assume that you know how to compile Camomile following the instructions on the repository and the demonstration is done for the effect plug-in only but this is exactly the same approach for the instrument version and the LV2 plug-in.

**Comment by emviveros** [comment]
I´m trying to compile Camomile with ELSE library in Windows, but I have a bunch of errors. I described all my procedure here: https://github.com/emviveros/Camomile-ELSE/issues/1

The error I has are in these file: 
[log_Camomile_ELSE_Effect_compilation_FAIL.txt](https://github.com/pierreguillot/Camomile/files/5633543/log_Camomile_ELSE_Effect_compilation_FAIL.txt)

I can compile ELSE with Camomile in OSx and Linux without problems.. Only Windows returns errors.

Thanks any advance!

**Comment by porres** [comment]
one of the errors is 
....\..\..\Source\Pd\ELSE\allpass.2nd~.c(49): error C2065: 'M_PI': undeclared identifier [C:\U
sers\Esteban\Documents\PortableGit\PortableGit\Camomile-ELSE\Effect\Builds\VisualStudio2015\
CamomileFx_SharedCode.vcxproj]

**M_PI** is from math.h and should be found. Seems like some linking issue

**Comment by pierreguillot** [comment]
https://stackoverflow.com/questions/26065359/m-pi-flagged-as-undeclared-identifier/26065595

**Comment by porres** [comment]
thanks, so maybe all the gazillion errors are not linking issues :(

**Comment by porres** [comment]
ok, here's a weird one! The error is 

..\..\..\Source\Pd\ELSE\keyboard.c(63): error C2057: expected constant expression [C:\Users\
Esteban\Documents\PortableGit\PortableGit\Camomile-ELSE\Effect\Builds\VisualStudio2015\Camom
ileFx_SharedCode.vcxproj]
..\..\..\Source\Pd\ELSE\keyboard.c(63): error C2466: cannot allocate an array of constant si
ze 0 [C:\Users\Esteban\Documents\PortableGit\PortableGit\Camomile-ELSE\Effect\Builds\VisualS
tudio2015\CamomileFx_SharedCode.vcxproj]
..\..\..\Source\Pd\ELSE\keyboard.c(63): error C2133: 'at': unknown size [C:\Users\Esteban\Do
cuments\PortableGit\PortableGit\Camomile-ELSE\Effect\Builds\VisualStudio2015\CamomileFx_Shar
edCode.vcxproj]

So it complains about line 64 in **keyboard.c** where it says it **_cannot allocate an array of constant si
ze 0_** and also complains about **_'at': unknown size_**

here's that line
 https://github.com/porres/pd-else/blob/ec6e2bda4f30198b0ba6f821885356d41f485eac/Classes/Source/keyboard.c#L63

basically, what we have is 

    int ac = 2;
    t_atom at[ac];

And it is complaining at has a size of 0 or unkown size, which doesn't seem to be true. How to fix that?

**Comment by porres** [comment]
it seems that if I just put an actual number and initialize it as t_atom at[2] it'll work, but I don't see why it can't get the value of the defined int in it... and although this is an easy fix for this object, there are others that I can't just hardcode a number into it, for instance: https://github.com/porres/pd-else/blob/aac57ae0374feb356a2e4743f18adbba8c42db3a/Classes/Source/note.c#L104

**Comment by pierreguillot** [comment]
https://stackoverflow.com/questions/9881777/why-do-i-get-cannot-allocate-an-array-of-constant-size-0
But the questions are really related to your library and it's more about standard coding in C. Please create an issue in your repository and tag me if necessary.

**Comment by pierreguillot** [comment]
As it seems that I will never find a direct and simple way to load externals into Camomile I started a [discussion](https://github.com/pierreguillot/Camomile/discussions/233) that present a workaround that IMO is very convenient. And I close this issue.
